### PR TITLE
chplcheck: warn for unused pattern matching like `(_, _)`

### PR DESCRIPTION
**Pull request by DanilaFe**
Fixes https://github.com/chapel-lang/chapel/issues/24838.

I believe it's better to warn about unused pattern matching separately from unused loop indices (one might avoid one but not the other), so this PR adds a new rule. This rule has an auto-fixit so you can turn unused `(_, _)` into `_` automatically.

Reviewed by @jabraham17 -- thanks!

## Testing
- [x] dyno tests
- [x] `test/chplcheck`

### CODE REVIEW

File: tools/chplcheck/src/chplcheck.py [hunk 1]
```diff
@@ -137,9 +137,18 @@ def apply_edits(edits: List[Edit], suffix: Optional[str]):
...
+
+        prev_start = None
         for edit in edits:
-            (line_start, char_start) = edit.start
-            (line_end, char_end) = edit.end
+            start = edit.start
+            line_start, char_start = edit.start
+            end = edit.end
+            line_end, char_end = edit.end
+
```
Comment (jabraham17) [inline tools/chplcheck/src/chplcheck.py:147]: The duplicated unwrapping of `edit.start` and `edit.end` is unfortunate, maybe replace the uses of `line_start` with `start[0]`, etc?

File: tools/chplcheck/src/rules.py [hunk 2]
```diff
@@ -26,6 +26,40 @@
...
+def variables(node: AstNode):
+    if isinstance(node, Variable):
+        if node.name() != "_":
+            yield node
+    elif isinstance(node, TupleDecl):
+        for child in node:
+            yield from variables(child)
+
+def file_lines_for_node(context: Context, node: AstNode):
+     return context.get_file_text(node.location().path()).split("\n")
```
Comment (jabraham17) [inline tools/chplcheck/src/rules.py:38]: remove this in favor of `get_file_lines` (which is in `chapel`)

File: tools/chplcheck/src/rules.py [hunk 3]
```diff
@@ -26,6 +26,40 @@
...
+
+def fixit_remove_unused_node(node: AstNode, lines: Optional[List[str]] = None, context: Optional[Context] = None):
```
Comment (jabraham17) [inline tools/chplcheck/src/rules.py:40]: nit, use an explicit return type

File: tools/chplcheck/src/rules.py [hunk 4]
```diff
@@ -26,6 +26,40 @@
...
+    parent = node.parent()
+    if parent is None:
+        return
```
Comment (jabraham17) [inline tools/chplcheck/src/rules.py:43]: nit, use an explicit `return None`

File: tools/chplcheck/src/rules.py (hunk 3 above)
Comment (jabraham17) [inline tools/chplcheck/src/rules.py:40]: maybe a doc comment here describing that `lines` or `context` must be supplied, and what `node` types this works on (iterand of a IndexableLoop and TupleDecl)

**Review by jabraham17 — APPROVED** [review]

File: tools/chplcheck/src/chplcheck.py (hunk 1 above)
Comment (DanilaFe) [reply tools/chplcheck/src/chplcheck.py:147]: I think there's too many uses of `start[0]` to make that reasonable. I'd rather unpack _and_ keep the top-level variable.

Comment (jabraham17) [reply tools/chplcheck/src/chplcheck.py:147]: In that case, I would replace uses  of `start` with `edit.start`. I see only 1 usage. Same with `end`

**Pull request merged**